*Note. This is my third Python Data Science project. The aim of this project was to become more comfortable with the `pandas` and `NumPy` libraries as data exploration and data cleaning tools and therefore limited analysis is undertaken. There are no data visualisations in this project as at the time of completing this project, the `Matplotlib` and `Seaborn` libraries were the next lessons in my data analytics course. This project will be revisited for further analysis and to bring the analysis to life with charts. Thank you for taking a look!*

# What are the drivers of second hand car prices

In this project, we will be exploring a dataset of used cars from eBay Kleinanzeigen, a classified section of the German eBay website. We will clean the data to prepare it for analysis to find potential drivers of price. 

There are are few individuals or agents involved in the sale or purchase of second hand vehicles that can benefit from  such analyses. Firstly, indivuduals who are looking to sell their cars can benefit from understanding this market by looking at how the car they are about to list is priced on avarage and what they can do to make their listing more attractive than their 'competitors'. Secondly, car buyers can also benefit from analysing this type of data as understanding the wider market, having information on average prices for the types of car they are looking for and the alternative listings that are available may provide them with additional bargaining power to secure the best deal they can. Lastly, agents and brokers that connect buyers and sellers or are resellers themselves (e.g. websites like [carwow](https://www.carwow.co.uk/), garages, private individuals) may also benefit from analysing this data by understanding how they can price cars to increase their margins or volume of sales. These can be the result of buying cars at the best price, pricing well when selling, or adding value to clients by giving them access to the information they are looking for.

From a learning perspective, the goal for this project is to become more proficient at using the `Pandas`and `NumPy` libraries in order to:
1. Explore data
2. Make informed decisions about how to clean data
3. Create new data objects for analysis and recommendations

## The Data

The data used for this project was originally scraped and uploaded to Kaggle by orgesleka. It can be downloaded from this [link](https://data.world/data-society/used-cars-data). The dataset contains data about listings of second hand cars on eBay Kleinanzeigen, a classified section of the German eBay website.

Here are the column names and their corresponding descriptions in eye-friendly format.

| Column Name in Dataset    | Description |
| :---------     | ----------: |
| "dateCrawled"   | when this ad was first crawled, all field-values are taken from this date |
| "name"   | "name" of the car |
| "seller"   | private or dealer |
| "offerType"   | type of listing: "offer" or "request" |
| "price"   | the price on the ad to sell the car |
| "abtest"   | whether the listing is included in an A/B test |
| "vehicleType"   | the vehicle Type |
| "yearOfRegistration"   | at which year the car was first registered |
| "gearbox"   | the transmission type |
| "powerPS"   | power of the car in PS |
| "model"   | the car model name |
| "kilometer"   | how many kilometers the car has driven |
| "monthOfRegistration"   | at which month the car was first registered |
| "fuelType"   | what type of fuel the car uses |
| "brand"   | the brand of the car |
| "notRepairedDamage"   | if the car has damage which is not repaired yet |
| "dateCreated"   | the date for which the ad at ebay was created |
| "nrOfPictures"   | number of pictures in the ad |
| "postalCode"   | the postal code for the location of the vehicle |
| "lastSeenOnline"   | when the crawler saw this ad last online |

We will determine the columns of interest for our analysis in the data exploration phase.

---

In [1]:
#importing the pandas and NumPy libraries
import pandas as pd
import numpy as np

## Exploring the Data

In [2]:
#using pd.read_csv to open and convert our csv file into a df
#setting encoding='Latin-1' due to encoding error with default 'UTF-8' encoding

autos = pd.read_csv('autos.csv', encoding = 'Latin-1', infer_datetime_format=True)

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
print (autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

### Primary Observations

The dataset contains: 50,000 rows with 20 columns.

There is a limited amount of columns containing non-null values:
* at most is `notRepairedDamage` with just under 20% non-null values.

Column names are not in a consistent format: most are in camelcase. Changing them to snakecase will make it easier to work with the data and provide consistency to the column names. 

Let's print the first 5 rows to see how the data appears. 

In [5]:
print (autos.head())

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

A recap of our first observations about the dataset itself:
* dataset size: 50,000 rows with 20 columns.
* non-null values not too problematic
* column names not in reader-friendly format: currently in camelcase

This dataset is in German, but as an English-speakers we can make assumptions about what the column descriptors refer to such as in the `gearbox` column. There are columns containing more technical words such as in `vehicle_Type`. It would be recommended to perform a quick translation before data analysis if this variable was analysed in more depth.

At first glance in the first 5 rows, we can make some observations about the quality of data format and data cleaning requirements. 

Here is a list of the columns which we could use in our analysis without performing significant data cleaning:
* `seller`
* `offerType`
* `abtest`
* `vehicleType
* `yearOfRegistration`
* `gearbox`
* `powerPS`
* `model`
* `fuelType`
* `brand
* `notRepairedDamage` is interesting as there is likely a difference of opinion on what is considered 'damage' by sellers and users and would definitely be a point of conversation between the two parties. It is likely a driver of price as damaged cars would most probably list for a lower price.
* `nrOfPictures` may not provide so much insight for analysis except as a proxy of transparency. Perhaps in an analysis where we look at cars that are sold (for which we would need sales data), this column could prove insightful.
* `postalCode`

Here are the columns of interest that will require more significant data cleaning:

* `price` (has a currency sign, may need further manipulation to convert to one currency for the entire dataset)
* `odometer` - distance driven is in km, has the `70,000km` format which needs cleaning and converting to numerical values before analysis

Lastly, we can see that the `name` column has amount of unique values, or ad names.  At first it may be logical to ignore this column for the moment as it will be resource intensive to clean. Some of the information in the ad `name` column is  provided in the other columns, for example the car brand in `brand`. There are however some interesting 'props' described in the `name` column such as in the entry `Peugeot_807_160_NAVTECH_ON_BOARD`. The information about the car having satellite navigation is not captured in any other column. This information which is not captured may explain variations in price between cars with similar attributes such as another Peugeot 807 without satelite navigation being cheaper. In the context of this project, we will not look so granularly at the data, but these observations should be noted as someone with the ability to clean that column in an efficient way could extract valuable insights.  

---

## Data Cleaning

### Data Cleaning - Renaming Columns

Let's begin by converting the column titles to snakecase and reword them to make them more descriptive.

In [6]:
autos.columns #retrieving column names

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Names to be changed from:to
``` 
    'dateCrawled' : 'date_crawled'
    'offerType' : 'offer_type'
    'vehicleType' : 'vehicle_type'
    'yearOfRegistration' : 'registration_year'
    'powerPS' : 'power_ps'
    'odometer' : 'odometer_km'
    'monthOfRegistration' : 'registration_month'
    'fuelType' : 'fuel_type'
    'notRepariedDamage' : 'unrepaired_damage'
    'dateCreated': 'ad_created'
    'nrOfPictures' : 'pictures'
    'postalCode' : 'postal_code'
    'last_seen' : 'last_seen'
```

In [7]:
#changing column names by hardcoding them over the previous labels
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures', 'postal_code',
       'last_seen']
autos.columns #confirming the columns have the new labels that we assigned

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures', 'postal_code',
       'last_seen'],
      dtype='object')

We made these changes in order to have a consistent format and make them easier to interepret and use while working with the `pandas` and `NumPy` libraries. 

### Data Cleaning - Removing Unnecessary Data

For our next phase of data cleaning, we will take a closer look at all the columns in the dataset.

In [8]:
autos.describe(include='all') #set the include parameter to "include='all'" to display non-numeric columns

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

       odometer_km  registration_month fuel_type       brand  \
count        50000        50000.000000     45518       50000   
unique          13                 NaN         7          40   
top      150,000km                 NaN    benzin  volkswagen   
freq         32424                 NaN     30107       10687   
mean           NaN            5.723360       NaN         NaN   
std            NaN            3.711984       NaN         NaN   
min            NaN            0.000000       NaN         NaN   
25%            NaN            3.000000       NaN         NaN   
50%            NaN            6.000000       NaN         NaN   
75%            NaN            9.000000       NaN         NaN   
max            NaN           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  pictures   postal_code  \
count              40171                50000   50000.0  50000.000000   
unique                 2                   76       NaN           NaN   
top                 nein  2016-04-03 00:00:00       NaN           NaN   
freq               35232                 1946       NaN           NaN   
mean                 NaN                  NaN       0.0  50813.627300   
std                  NaN                  NaN       0.0  25779.747957   
min                  NaN                  NaN       0.0   1067.000000   
25%                  NaN                  NaN       0.0  30451.000000   
50%                  NaN                  NaN       0.0  49577.000000   
75%                  NaN                  NaN       0.0  71540.000000   
max                  NaN                  NaN       0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

We can observe from this table that `seller` and `offer_type` columns have `49999` counts of the same type of entry. We can remove them from our dataset as they won't provide value as potential analytical cuts. 

As mentioned previously, the `pictures` could have provided interesting 'proxy' information, but we can infer from the percentiles figures that there are likely no pictures in any of the 50,000 listings in this dataset. 

We will need to convert the `price` and `odometer_km` columns into numerical ones in order to be able to perform calculations in our analysis phase.

The `registration_year` column will require data cleaning as its `min` is `1000` and `max` is `9999`, which is incorrect as cars didn't exist in the year 1000, nor can data be about the future.

In [9]:
autos['pictures'].value_counts()

0    50000
Name: pictures, dtype: int64

This column clearly has no valuable information and we will therefore drop it. 

In [10]:
autos = autos.drop(['seller','offer_type','pictures'],axis = 1) #removing the three columns 

Let's clean the `price` column and turn it into a number.

In [11]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int) 
autos['price'].describe()

/var/folders/y8/wqz8h72s2xj544lcdblf8q000000gq/T/ipykernel_64119/2251124695.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

We have now removed the `$` sign as well as the `,000s` comma separator. The column was converted to `float` type to be able to perform calculations. 

We can observe that the max `price` is `1.000000e+08` which indicates a car price over `100,000,000` 
which is not realistic. 

We can observe that the `min` value for `price` is `0`. This should not be the case as it is unlikely a legitimate seller would be selling their car for a below or equal to `0`.

The above two observations signal that we have outliers and need to remove them from the dataset before we can proceed to the analysis phase. 

### Data Cleaning - Removing outliers in the `price` column

We will begin by looking at the top end of the `price` column. We will consider any value above `72000` high as it is where the 75th percentile begins. 

In [12]:
high_price = autos[autos['price']> 72000].sort_values('price',ascending=False) #sort values by descending order
high_price.head(30) #checking the top 30 rows by highest value in 'price'

date_crawled                                               name  \
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
36818  2016-03-27 18:37:37                                        Porsche_991   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
47337  2016-04-05 10:25:38                                    BMW_Z8_roadster   
38299  2016-03-28 22:25:25                                Glas_BMW_mit_Wasser   
37840  2016-03-21 10:50:12                                        Porsche_997   
40918  2016-03-20 18:40:05                             Porsche_911_991_GT3_RS   
43668  2016-03-16 18:47:26  Porsche_993/911_Turbo_WLS_II_Exclusive_S_deuts...   
28090  2016-03-08 19:50:06  Tesla_Model_X_P90D_Signature_Sondermodel__Neuw...   
20351  2016-03-22 17:47:59                                    BMW_Z8_roadster   
17140  2016-03-05 17:39:22  Mercedes_Benz_SLS_AMG_Roadster_AMG_SPEEDSHIFT_DCT   
11433  2016-03-30 21:44:21  Audi_R8_V10_plus_5.2_FSI_quattro_S_tronic_Akra...   
32840  2016-03-22 08:55:27  Mercedes_Benz_G_500_7G_TRONIC_Cabrio_Desingno_...   
18509  2016-03-27 18:43:02                                    Porsche_991_GT3   

          price   abtest vehicle_type  registration_year    gearbox  power_ps  \
39705  99999999  control    limousine               1999  automatik       224   
42221  27322222  control    limousine               2014    manuell       163   
39377  12345678  control          NaN               2018    manuell        95   
47598  12345678  control    limousine               2001    manuell       101   
27371  12345678  control          NaN               2017        NaN        95   
24384  11111111     test          NaN               1995        NaN         0   
2897   11111111     test    limousine               1973    manuell        48   
11137  10000000  control        coupe               1960    manuell       368   
47634   3890000     test        coupe               2006        NaN       799   
7814    1300000  control        coupe               1992        NaN         0   
22947   1234566  control        kombi               1999  automatik       190   
514      999999     test        kombi               2009    manuell       101   
43049    999999     test          bus               1981    manuell        70   
37585    999990     test    limousine               1985    manuell       111   
36818    350000  control        coupe               2016    manuell       500   
14715    345000  control       cabrio               2012  automatik       460   
34723    299000     test        coupe               1977    manuell       260   
35923    295000     test       cabrio       

The table above shows the first 30 entries in descending order with a `price > 72000` which is where our 75th percentile begins. We can notice anomalies in the top 10 rows when comparing the `price` to the corresponding car description. It can be expected that a `Ferrari_FXX` can have a price of `3,890,000` $ [(price data)](https://www.classic.com/m/ferrari/fxx/) (though suprising to see it listed on eBay!). The cars above the `Ferrari_FXX` are unlikely to be accurate entries, as are the four rows that follow it (if having a little knowledge about the car market).

To remove the 12 outlier listings we will:
* Create a boolean mask to remove values above `3,890,000` which is the price of the `Ferrari_FXX` - the accurately top priced car in this list 
* Use the `df.drop()` function to remove the rows we know by their index number.

We will create an `autos_clean` dataset in which we will store cleaned data from now on.

In [13]:
autos_clean = autos[autos['price'] < 3890001] #creating our new dataset and using the boolean mask to remove prices above 3890001
autos_clean.shape #check to see if we have removed the values above 3890001

(49992, 17)

`49992` rows confirms we have removed the 8 rows above the `Ferrari_FXX` listing.

We'll now use the `df.drop()` function.

In [14]:
#we are dropping the 4 values we know the index by row of - [22947,514,43049,37585]
#axis = 0 will search for index number in rows
#inplace=True ensures we replace DataFrame without having to apply this mask back to autos_clean

autos_clean.drop([22947,514,43049,37585], axis = 0, inplace=True)

/var/folders/y8/wqz8h72s2xj544lcdblf8q000000gq/T/ipykernel_64119/1843654163.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos_clean.drop([22947,514,43049,37585], axis = 0, inplace=True)


In [15]:
autos_clean.shape

(49988, 17)

We can confirm 12 rows were removed from the original dataset.

Let's turn out attention to the bottom end of the price range.

In [16]:
low_price = autos[autos['price'] == 0]
print(low_price.shape) #will show us the amount of ads with a listing price of 0$
low_price_100 = low_price.head(100) #will show us if the entries with listing price of 0$ have any other missing information
low_price_100

(1421, 17)


date_crawled                                               name  \
27    2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
71    2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80    2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87    2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99    2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
...                   ...                                                ...   
3115  2016-03-15 00:57:52                        BMW_E36_320i_Cabrio_M_Paket   
3169  2016-03-18 20:44:43                           Tausch_Audi_a4_b5_1.8_5v   
3178  2016-03-07 20:51:11      LACKIERARBEITEN_vom_Profi_zum_FAIREN_Preis!!!   
3248  2016-03-17 13:49:06                              Renault_Megane_Scenic   
3308  2016-03-28 20:58:50                      Suche_dringend_ein_Kleinwagen   

      price   abtest vehicle_type  registration_year  gearbox  power_ps  \
27        0  control          NaN               2005      NaN         0   
71        0  control          NaN               1990  manuell         0   
80        0  control        coupe               1999  manuell        99   
87        0  control          NaN               2000      NaN         0   
99        0  control       cabrio               2011  manuell         0   
...     ...      ...          ...                ...      ...       ...   
3115      0     test       cabrio               1998  manuell       150   
3169      0     test    limousine               1998  manuell       125   
3178      0     test          NaN               1990      NaN         0   
3248      0  control          bus               2004  manuell         0   
3308      0  control          NaN               2000      NaN         0   

        model odometer_km  registration_month fuel_type           brand  \
27        NaN   150,000km                   0       NaN            ford   
71        NaN     5,000km                   0    benzin            opel   
80    primera   150,000km                   3    benzin          nissan   
87        5er   150,000km                   0       NaN             bmw   
99    2_reihe    60,000km                   7    diesel         peugeot   
...       ...         ...                 ...       ...             ...   
3115      NaN   150,000km                   0       NaN             bmw   
3169       a4   150,000km                   3    benzin            audi   
3178      NaN   150,000km                   0       NaN         porsche   
3248   megane   150,000km                   8    diesel         renault   
3308      NaN     5,000km                   0       NaN  sonstige_autos   

     unrepaired_damage           ad_created  postal_code            last_seen  
27                 NaN  2016-03-27 00:00:00        66701  2016-03-27 18:45:01  
71                 NaN  2016-03-28 00:00:00         4552  2016-04-07 01:45:48  
80                  ja  2016-03-09 00:00:00        66903  2016-03-09 16:43:50  
87                 NaN  2016-03-29 00:00:00        82256  2016-04-06 21:18:15  
99                nein  2016-04-05 00:00:00        99735  2016-04-07 12:17:34  
...                ...                  ...          ...                  ...  
3115                ja  2016-03-15 00:00:00        56729  2016-03-18 00:47:19  
3169               NaN  2016-03-18 00:00:00        51545  2016-04-05 22:47:19  
3178               NaN  2016-03-07 00:00:00        71636  2016-03-08 02:16:48  
3248                ja  2016-03-17 00:00:00        72213  2016-03-17 13:49:06  
3308               NaN  2016-03-28 00:00:00        44532  2016-04-03 02:20:29  

[100 rows x 17 columns]

There are `1,421` car ads with a listing price of `0`, just above 2.8% of all entries in the dataset. 

While the ads with a listing price of `0` may containt data in other columns, the rows with `0` in `price` may skew our data due to being outliers so we will remove them before we proceed to the data analysis phase. 

Let's check for entries with a listing price between `0` and `300` and also treat them as outliers. 

While this data was originally collected in 2016, it is unlikely people would be selling a car at a price below `300` even being second hand ones. 

In [17]:
low_price = autos[autos['price'] < 300]
low_price.shape

(3118, 17)

There are `3,118` car adverts with a listing price below `300`. This represents just over 6.2% of all entries in the dataset.

To confirm that we can remove this data from the dataset, let's have a look at the `registration_year` as some really old cars may actually sell below `300`.

In [18]:
low_price['registration_year'].value_counts().sort_values(ascending=False).head(15)

2000    430
1997    288
1995    271
1998    254
1999    223
2016    219
1996    216
2005    124
2001    107
2017     97
1994     95
1990     91
2002     75
1993     62
1992     57
Name: registration_year, dtype: int64

The years from `1990` to `2017` are the 15 most frequent by `registration_year` for cars with a price below `300`. This indicates that low priced cars are in car listings with both recent and older car registration years. 

We can therefore remove this data for our data analysis phase as it demonstrates these price points are outliers. Let's resurface the `autos_clean` dataset we are building for our analysis phase. 

In [19]:
autos_clean = autos_clean[autos_clean['price'] > 299]
autos_clean.shape

(46870, 17)

The above confirms we removed `3130` rows in total to create our `autos_clean` dataset.
* `12` rows from high price outliers
* `3118` rows from low price outliers

The `data_clean` data set now contains just under 94% of the original data.

### Data Cleaning - Formatting the `odometer_km` column

Let's turn our attention to the `odometer_km` column. As a reminder, it tells us car mileage.

In [20]:
autos['odometer_km'].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer_km, dtype: int64

In order to be able to perform calculations with the data in the `odometer_km` column, we need to remove the '`,000s`' separator and the `km` at the end of each entry. This will allow us to turn the column into a numerical type `int` or `float`. 

In [21]:
autos_clean['odometer_km'] = autos_clean['odometer_km'].str.replace(',','').str.replace('km','').astype(int)
autos_clean['odometer_km'].value_counts().sort_index(ascending=False)

150000    30144
125000     4953
100000     2069
90000      1705
80000      1401
70000      1204
60000      1141
50000      1004
40000       812
30000       771
20000       736
10000       240
5000        690
Name: odometer_km, dtype: int64

We have removed the `,000s` separator and the `km` at the end of each entry in the `odometer_km` column.

Most of the car ads have a value of `150000` in the `odometer_km` column which makes sense as these are second hand cars, with each additional distance 'tranche' starting from `10000` becoming more and more frequent. The `5000` tranche would be nested between the `50000` and `60000` tranches if we displayed the series in order of frequency. 

Unlike in the `price` column, there do not appear to be outliers that could skew our data in `odometer_km`. 

### What is the date range within which these ads were posted?

We should consider the date range of these advertisements in order to contextualise our analysis to that time.

We will look at year in which the ads were posted.

In [22]:
date_range = autos_clean['ad_created']
date_range.describe()

count                   46870
unique                     76
top       2016-04-03 00:00:00
freq                     1822
Name: ad_created, dtype: object

In order to extract just the year, we will keep just the first four characters of the `string` that is the entry for that column.

In [23]:
date_range = autos_clean['ad_created'].str[:4]
print(date_range.unique())
print(date_range.value_counts(normalize=True))
print()

['2016' '2015']
2016    0.999872
2015    0.000128
Name: ad_created, dtype: float64



This dataset contains ads created in 2015 and 2016, with almost all of them created in the year 2016.

If we had a dataset with a more extensive time range and logs of changes in price (if applicable) per listing, we could look at how prices for similar cars change over time.

### Data Cleaning - Removing outliers in the  `registration_year` column

Let's take a look at the `registration_year` column. As a reminder it has a `min` value of `1000` and `max` of `9999` which are not correct values for true registration years. 

In [24]:
autos_clean['registration_year'].value_counts().sort_index()

1000    1
1001    1
1111    1
1800    2
1910    2
       ..
5911    1
6200    1
8888    1
9000    1
9999    3
Name: registration_year, Length: 95, dtype: int64

The data above shows us we should investigate both the top end and bottom end of the `registration_year` column to identify the range the outliers are in and remove them from our `autos_clean` dataset. 

In [25]:
autos_clean['registration_year'].value_counts().sort_index().head(15)

1000    1
1001    1
1111    1
1800    2
1910    2
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
Name: registration_year, dtype: int64

In [26]:
autos_clean[autos_clean['registration_year'] <1930]

date_crawled                                   name  price  \
10556  2016-04-01 06:02:10                             UNFAL_Auto    450   
21416  2016-03-12 08:36:21                Essex_super_six__Ford_A  16500   
22101  2016-03-09 16:51:17                             BMW_Andere  11500   
22316  2016-03-29 16:56:41  VW_Kaefer.__Zwei_zum_Preis_von_einem.   1500   
22659  2016-03-14 08:51:18                           Opel_Corsa_B    500   
24511  2016-03-17 19:45:11            Trabant__wartburg__Ostalgie    490   
28693  2016-03-22 17:48:41                         Renault_Twingo    599   
32585  2016-04-02 16:56:39                             UNFAL_Auto    450   
49283  2016-03-15 18:38:53                             Citroen_HY   7750   

        abtest vehicle_type  registration_year  gearbox  power_ps   model  \
10556  control          NaN               1800      NaN      1800     NaN   
21416  control       cabrio               1927  manuell        40  andere   
22101     test       cabrio               1929  manuell        15  andere   
22316  control          NaN               1000  manuell         0  kaefer   
22659     test          NaN               1910      NaN         0   corsa   
24511  control          NaN               1111      NaN         0     NaN   
28693  control   kleinwagen               1910  manuell         0     NaN   
32585  control          NaN               1800      NaN      1800     NaN   
49283  control          NaN               1001      NaN         0  andere   

       odometer_km  registration_month fuel_type       brand  \
10556         5000                   2       NaN  mitsubishi   
21416         5000                   5    benzin        ford   
22101         5000                   1       NaN         bmw   
22316         5000                   0    benzin  volkswagen   
22659       150000                   0       NaN        opel   
24511         5000                   0       NaN     trabant   
28693         5000                   0    benzin     renault   
32585         5000                   2       NaN  mitsubishi   
49283         5000                   0       NaN     citroen   

      unrepaired_damage           ad_created  postal_code            last_seen  
10556              nein  2016-04-01 00:00:00        63322  2016-04-01 09:42:30  
21416               NaN  2016-03-12 00:00:00        74821  2016-03-15 12:45:12  
22101                ja  2016-03-09 00:00:00        70569  2016-04-07 06:17:11  
22316               NaN  2016-03-29 00:00:00        48324  2016-03-31 10:15:28  
22659               NaN  2016-03-14 00:00:00        52393  2016-04-03 07:53:55  
24511               NaN  2016-03-17 00:00:00        16818  2016-04-07 07:17:29  
28693               NaN  2016-03-22 00:00:00        70376  2016-04-06 09:16:59  
32585              nein  2016-04-02 00:00:00        63322  2016-04-04 14:46:21  
49283               NaN  2016-03-15 00:00:00        66706  2016-04-06 18:47:20

We will remove rows with `registration_year` values below `1911` as they are outliers because the registration years cannot be the car brands and models being advertised - Opel Corsa and Renault Twingo are modern cars. On the other hand, the `Essex_super_six__Ford_A` is a car with possible registration year of `1927` - we can confirm at this [link](https://www.vintagecarcollector.com/vehicles/1898/1929-essex-super-six) of a vintage cars website.

In [27]:
autos_clean['registration_year'].value_counts().sort_index().tail(15)

2015     372
2016    1097
2017    1355
2018     466
2019       1
2800       1
4100       1
4500       1
4800       1
5000       2
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

We'll remove any rows with `registration_year` values above `2016`. These must be outliers or incorrect entries because cars cannot be registered in the future and we know that the listings were created in 2015 and 2016. 

In [28]:
#only keeping entries with a 'registration_year' value between 1911 and 2016

autos_clean = autos_clean[autos_clean["registration_year"].between(1911,2016)] 
autos_clean.shape

(45028, 17)

We have removed a further 1,842 rows in the data cleaning process of the `registration_year` outliers and incorrect entries.

In [29]:
autos_clean['registration_year'].describe()

count    45028.000000
mean      2003.072177
std          7.069710
min       1927.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [30]:
autos_clean['registration_year'].value_counts(normalize=True,dropna=False).head(10)

2000    0.064937
2005    0.064204
1999    0.061628
2004    0.059807
2003    0.059607
2006    0.059252
2001    0.057631
2002    0.054588
2007    0.050457
2008    0.048992
Name: registration_year, dtype: float64

In [31]:
sum(autos_clean['registration_year'].value_counts(normalize=True,dropna=False).head(9))

0.5321133516922802

When analysing the `registration_year` column in our clean dataset, we can observe that the most frequent years in wich these vehicles were registered are between 1999 and 2006 (frequency > 5%). Including 2007 (frequency < 5%) we can deduce that half the cars sold on eBay, have been driven for 9 to 18 years at the time they were listed. 

### Summary of Data Cleaning

Up to now we have:
* removed 3 columns (`seller`,`offer_type`,`pictures`) as they did not provide meaningful data for our analysis
* converted the `price` and `odometer_km` columns from integer to numerical to be able to perform calculations
* removed rows with outliers on the high and low end of the `price` and `registration_year` columns

During this phase we created an `autos_clean` dataset which contains 45,028 rows (just above 90% of our original data) and 17 columns that we will use in the analysis phase. 

---

## Data Analysis

For the data analysis, let's have a look at our dataset and see which columns might be interesting to analyse by aggregation. 

In [32]:
autos_clean.describe(include ='all')

date_crawled      name         price abtest vehicle_type  \
count                 45028     45028  4.502800e+04  45028        42810   
unique                43583     34409           NaN      2            8   
top     2016-03-05 16:57:05  BMW_316i           NaN   test    limousine   
freq                      3        75           NaN  23215        12369   
mean                    NaN       NaN  6.306413e+03    NaN          NaN   
std                     NaN       NaN  2.140559e+04    NaN          NaN   
min                     NaN       NaN  3.000000e+02    NaN          NaN   
25%                     NaN       NaN  1.399000e+03    NaN          NaN   
50%                     NaN       NaN  3.300000e+03    NaN          NaN   
75%                     NaN       NaN  7.800000e+03    NaN          NaN   
max                     NaN       NaN  3.890000e+06    NaN          NaN   

        registration_year  gearbox      power_ps  model    odometer_km  \
count        45028.000000    43251  45028.000000  43041   45028.000000   
unique                NaN        2           NaN    243            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    33490           NaN   3579            NaN   
mean          2003.072177      NaN    120.233766    NaN  125516.012259   
std              7.069710      NaN    186.769909    NaN   39586.648614   
min           1927.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     75.000000    NaN  100000.000000   
50%           2003.000000      NaN    110.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           2016.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         45028.000000     42164       45028             37546   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     27410        9511             33492   
mean              5.899018       NaN         NaN               NaN   
std               3.637681       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 45028  45028.000000                45028  
unique                   74           NaN                35943  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1758           NaN                    8  
mean                    NaN  51249.690837                  NaN  
std                     NaN  25718.715258                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30916.000000                  NaN  
50%                     NaN  50181.000000                  NaN  
75%                     NaN  72138.000000                  NaN  
max                     NaN  99998.000000                  NaN

There are lots of interesting things to explore. 

If we think about what car 'attributes' individuals consider when selling or buying a car and how that ultimately affects price, these usually would be:
* car type (depending on need: eg. family car, van, sports car)
* brand & model
* distance driven
* registration year
* damage.

They may also consider:
* fuel type 
* gearbox (some individuals may prefer to drive automatic)
* power.

Ultimately, making a car sale or purchase involves the weighing of these 'attributes' and the seller or buyer's weighed importance in each.

Let us explore the car brand and how that affects price as it is a popular albeit one-dimensional way of looking at price. 

In [33]:
autos_clean['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'renault', 'mercedes_benz', 'audi', 'sonstige_autos', 'opel',
       'mazda', 'porsche', 'mini', 'seat', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [34]:
autos_clean['brand'].value_counts(normalize=True,dropna=False).head(15)

volkswagen       0.211224
bmw              0.112397
opel             0.104202
mercedes_benz    0.099049
audi             0.088678
ford             0.067314
renault          0.045705
peugeot          0.029626
fiat             0.024762
seat             0.018189
skoda            0.016745
mazda            0.015191
nissan           0.015124
smart            0.014613
citroen          0.014102
Name: brand, dtype: float64

Unsurprizingly the top 5 most frequent brands in the car listings are German. This is a dataset about second cars on  German eBay after all. Let's have a look at at the top 10 brands. Today Seat is part of the Wolkswagen Group but for the purposes of this exercise we will treat them as separate.

In [35]:
top_brands = autos_clean['brand'].value_counts(normalize=True,dropna=False).head(10)
top_brands.index #creating the index for our list of top brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')

We will first calculate the mean price for a single brand - Volkswagen.

Our calculation will give us a mean price value for all vehicle types. We know that there is a `vehicle_type` column that could make this calculation more insightful, such as comparing how types of vehicles vary by price accross brands.

In [36]:
vw_price = autos_clean[autos_clean['brand']=='volkswagen']['price']
vw_mean = vw_price.sum()/vw_price.size # dividing the sum of the series by the amount of items in the series
#or
vw_mean = vw_price.mean()

print(vw_mean)

5595.6266428346125


We will need to loop this process for each `brand` in our `top_brands.index`.

In [37]:
#creating dictionary to hold the aggregate data
brand_price = {} 

#using a for loop to create the mean price data for each brand
for oem in top_brands.index: #oem refers to original equipment manufacturer
    brand = autos_clean[autos_clean['brand']== oem]
    price = brand['price']
    mean_price = price.mean() 
    brand_price[oem] = int(mean_price) #appending our dictionary with brand name as key, and average price as an integer value

brand_price

{'volkswagen': 5595,
 'bmw': 8455,
 'opel': 3172,
 'mercedes_benz': 8710,
 'audi': 9447,
 'ford': 4022,
 'renault': 2634,
 'peugeot': 3222,
 'fiat': 3008,
 'seat': 4573}

### Observations

Audi, Mercedez-Benz, and BMW are on average the most expensive brands in this dataset. Volkswagen and Seat can be considered mid-range priced vehicles. The remaining brands (Ford, Peugeot, Opel, Fiat, and Renault) have an average price below `4,000` and are generally cheaper. These observations are somewhat expected as the top 3 brands are considered more premium than the other brands in our list.

We suggested earlier there are several 'attributes' that individuals will weigh and consider when buying and selling second hand vehicles. We could hypothesize that brand (premium, mid-range, budget range) is positively correlated with car price - brands with higher brand value are priced higher.

The analysis above shows brand name can be one driver of price, with top brands being the highest-priced cars on average. We will also have a look at the mileage (`odometer_km` column) for the brands. 

While we calculate the mean mileage for each brand, a remember that the mileage (`odometer_km`) column is categorical and provided mileage data in tranches. This means the series will not demonstrate the true average mileage of the cars but we can use the given data as a proxy. 

In [38]:
#creating dictionary to hold the aggregate data
brand_mileage = {} 

#using a for loop to create the mean mileage data for each brand
for oem in top_brands.index: #oem refers to original equipment manufacturer
    brand = autos_clean[autos_clean['brand']== oem]
    mileage = brand['odometer_km']
    mean_mileage = mileage.mean() 
    brand_mileage[oem] = int(mean_mileage) #appending our dictionary with brand name as key, and average mileage as an integer value

brand_mileage

{'volkswagen': 128570,
 'bmw': 132763,
 'opel': 128919,
 'mercedes_benz': 131114,
 'audi': 129189,
 'ford': 123927,
 'renault': 127480,
 'peugeot': 126604,
 'fiat': 116412,
 'seat': 121428}

In [39]:
#creating price and mileage series from our dictionary
brand_mean_p = pd.Series(brand_price)  
brand_mean_m = pd.Series(brand_mileage)

#creating the price and mileage DataFrames
df_bmp = pd.DataFrame(brand_mean_p, columns=['mean_price']) 
df_bmm = pd.DataFrame(brand_mean_m, columns=['mean_mileage_km'])

#merging the datframes together using the rows as indices
df_merged = pd.merge(df_bmp,df_bmm,right_index=True,left_index=True) 
df_merged.sort_values('mean_price',axis=0,ascending=False) #sort it by 'mean_price'

mean_price  mean_mileage_km
audi                 9447           129189
mercedes_benz        8710           131114
bmw                  8455           132763
volkswagen           5595           128570
seat                 4573           121428
ford                 4022           123927
peugeot              3222           126604
opel                 3172           128919
fiat                 3008           116412
renault              2634           127480

### Final Observations

It can be observed that there isn't much variation in the average mileage acrooss the cars - 8 out of 10 in this list  have average mileage between 122,000km and 133,000km. 

Noticeably, the top 3 brands by price: Audi, BMW, and Mercedez-Benz have the highest average mileage. It is compelling that despite the highest mileage on average remain most expensive on average in the data. 

We can hypothesize that despite a higher mileage on average, sellers know that buyers are willing to pay a higher price for 'premium' brands. A deeper analysis into the price of cars for each brand by tranche of mileage, for example: 50,000km and above, 75,000km and above, and so on could reveal interesting findings. A possible avenue to explore is how price decreases (if it does) with mileage across different brand types, from premium to lower end cars. We can hypothesize that more premium cars see a relatively lower percentage price decrease with increased mileage than lower end cars.

---

## Conclusion

In conclusion we found that brand can be a driver of price, with premium brands listing for a higher price. Interestingly, the premium brands also had the highest mileage on average. We came to these findings by analysing the listings of the top 10 most frequent car brands in this dataset.

These initial findings open the floor for interesting analyses on the drivers of second hand car prices. The insights about car prices could be of relevance for today's second hand car marketplaces and brokers as it is a growing market (due to chip shortages) at the time of writing. 

The ads in this dataset were created in 2015 and 2016.While the psychology behind selling or buying cars may not have changed much since then, trends in the industry and changes in circumstances have likely affected the prices which may lead to different observations and conslusions. Furthermore, this dataset is limited to 50,000 datapoints (45,000 used in the analysis) which would paint an incomplete picture of the second hand car market even if contextualized to that time. 

From a learning perspective, it is evident that the `Pandas` and `NumPy` libraries are great tools for initial data exploration, cleaning, and for analysis as well. The inbuilt libraries and Series and DataFrame formats are much more efficient to work with than lists and dictionaries to create sets of data to be analysed. One of the key advantages is to be able to use column names in the functions and methods, making it easier to work with as well as easier to follow for the reader. This project has made me appreciate the power of libraries and the people who take time to develop and expand them.